### ISSUE TITLE
[ARM/Linux] Make indirect calls to direct calls if possible

### ISSUE THREAD

**Issue by wateret** [dotnet/coreclr#7002]
In this issue I'd define direct/indirect call as jump to pc-relative immediate address/jump to an absolute address from a register.

**Direct**
```
bl imm24
```

**Indirect**
```
movw reg, imm16
movt reg, imm16
blx reg
```

On ARM32, it seems like indirect calls are generated for almost all cases even if the caller and callee address are close enough to fit in imm24. `CEEJitInfo::getRelocTypeHint()` always returns *no-hint* which results in an indirect call.

I'd like to calculate the relative address between the address of current instruction and the target address then we are able to decide whether to use direct or indirect. However it doesn't look like we are able to know the address of current instruction at the time cause it is before generating actual instructions. Is there some ways to handle this?

/cc @dotnet/jit-contrib @hqueue

**Comment by wateret** [comment]
I tested this for recursive calls, Code is here: https://github.com/wateret/coreclr/commit/215f6405bb8616e03bc3b74c60f9d47b5d0d30ca

Then I ran some recursion-dominated benchmark tests.

| TestCase         | Vanilla | Patched | Vanilla/Patched |
|------------------|---------|---------|-----------------|
| BenchI-Ackermann | 4829    | 4431    | 1.09            |
| BenchI-Fib       | 5537    | 4733    | 1.17            |
(unit: ms)

**Comment by briansull** [comment]
Yes, the JIT doesn't know the address for the current PC,  because we haven't yet asked the VM for that memory.  It is a problem that is somewhat difficult to solve for the general case.

**Comment by DemiMarie** [comment]
@briansull Could the JIT ask for the memory from the VM and then patch the instructions later?

**Comment by wateret** [comment]
@DemiMarie I think his comment here would be the answer. https://github.com/dotnet/coreclr/pull/7310#issuecomment-249945594

**Comment by briansull** [comment]
We could extend the JIT to EE interface so that we could pre-request a code buffer of a given size.
 The function could take a code-size and return a tuple of [request-cookie, code-address, code-size]
 You could then be guaranteed a code buffer of code-size at a fixed address that you could rely upon.
 Then the codegen phase could either keep and use this code-buffer or abandon it.

**Comment by BruceForstall** [comment]
Also: https://github.com/dotnet/coreclr/issues/19182

On some other architectures, we assume the best, then if there is a problem, back off and rejit assuming the worst. We probably need something in between these two extremes.
